src/plan.rs

```diff
@@ -67,6 +73,7 @@ pub struct ShapePlanner<'a> {
     pub script_zero_marks: bool,
     pub script_fallback_mark_positioning: bool,
     pub shaper: &'static ComplexShaper,
+    pub user_features: Vec<Feature>,
```

Comment:
Why do we have to allocate it again?

Reply:
But `compile` is a private method. It cannot be misused.

As for allocation, yes, my bad. There is none. But the code is still confusing. It confused me. I would prefer to avoid storing `user_features` in the `ShapePlanner`.

Reply:
Sure, done.  I'm used to working on codebases with many more contributors (and PR approvers), where there's more value in protecting even private functions from misuse within the codebase. :)